### ISSUE TITLE
How decoupled is this architecture?

### ISSUE THREAD

**Issue by jackrabbithanna** [totten/afform#16]
I'm starting to think outside of the Drupal box, and I'm wanting to understand what's going on here. Is it possible that Mark moves past Drupal-only thinking, IDK ....Decouple so to speak?

I'm thinking about how to make CiviCRM form functionality be completely decoupled from the CMS, or even the PHP based CiviCRM system as it is now.

The documentation reads to me like you still need a CiviCRM extension, and these forms will require the CiviCRM PHP system ... but maybe I'm missing something. You have nice REST API, with new and better on the way .. so technically it's already possible to decouple CiviCRM. What's missing is some standardization where all a content editor needs to do is add a <civicrm-contribution-form></civicrm-conribution-form> tag

It seems to me that it could be possible to have a JS infrastructure where you can import the necessary JS, have your html templates, and drop that anywhere, and have the same form, with ajax to do the API requests to make a contribution, register for an event etc...

I'm thinking about functional units that could be placed in a static html website, Wordpress, Joomla, Drupal, served up by NodeJS, or even copy/pasted into something like Wix...Could this get to a place where no matter how you render the html and load the javascript, you get standardized CiviCRM functionality? How decoupled can this get?

I don't much about these javascript frameworks, there are many of them, but I'm willing to learn now and be open to new ideas, and I'm seeing that it's possible to some very amazing things...

This is the video on webcomponents really kicked down the walls in my mind: https://www.youtube.com/watch?v=Akn6keIYZ3k

https://www.webcomponents.org/

Same guy is building something called HAX (Headless Authoring Experience) https://haxtheweb.org/
Which works like that, you can drop it into Drupal, Wordpress, Joomla, static html, whatever...

Webcomponents are JS library agnostic, in the sense that they could be used with (or without) Angular or React or whatever. So as far as my limited understand goes, you can use webcomponents with AngularJS, it's not an either/or proposition...Theoretically, could could make webcomponents that are actually lil mini Angular apps.

I'm not trying to push any ideas for this project, because I'm not going to pretend I understand the cutting edge JS world. It's just this webcomponents standard has really opened me up to a different kind of thinking.

Does Afform have similarities / differences to webcomponents?

So just how decoupled is this Afform architecture? How far could it go?

I'm just thinking that there could be standard CiviCRM webcomponents or "whatever" where any framework, any CMS, could serve up standardized civicrm contribution / event registration / profile / any forms.

You guys are way out in the frontier here, and you're busy and don't need to explain it to me, but if you feel like it, share your thoughts with a reforming Drupal purist exploring what is to me a brave new JS world.

Anyway, I'm thinking about new things, and going to try this stuff out here, and see what it can be combined with and how it can be stretched and pulled, and of course fit into Drupal in new and unique ways :).

**Comment by JoeMurray** [comment]
Thanks for these thoughts and links. I sometimes try to imagine leapfrogging into the headless world too. I've been thinking it woupd require everything bespoke in js calling the API but perhaps not... Need to do some more research.

**Comment by jackrabbithanna** [comment]
yeah I'm imaging a situation where CiviCRM community could encapsulate form element / form / form action into webcomponents so that downstream people can just include them and get near instantaneous CiviCRM functionality, not just CMS independent, but just works everywhere ...Just thinking about exploring web components and making something useful Civi-related at the same time. Not only does that idea gel with a SaSS product, but it would be possible for other partners to do their own SaSS type thing with the same architecture. Thinking about orgs that have CiviCRM, but want to spin up nifty little sort of temporary marketing sites for individual events, or contribution campaigns, etc.. and not be locked in to any particular CMS or a CMS at all. Could totally work for partners to have their own CDN and push functionality to all clients with consistent look/feel/form action too ... I also think these things are so cool that I'd share the information about the technology and see what people think....So I'm wondering if the architecture for form builder / forms here can or could be detached from an installed CiviCRM instance .. seems like if only API is used, then it could be possible.  That is not needing to be in an extension or a page served up by Civi / the CMS. As I understand it, it may be possible to wrap up AngularJS apps in a web component. For sure it is possible to use web components in an AngularJS app...
